### ISSUE TITLE
Provide a way to AOT torch.compile and serialize a model

### ISSUE THREAD

**Issue by pritamdamania87** [pytorch/pytorch#113287]
### 🚀 The feature, motivation and pitch

For large scale training jobs, it is very common to run the same model architecture multiple times across several jobs. For models which can be trained using just DataParallel training, there is a large amount time spent just compiling the model before actual training begins.

For N gpus where N is large, this is quite wasteful and expensive since all the GPUs are doing exactly the same work and compiling the exact same model on exactly same input shapes/types.

Note that despite using `TORCHINDUCTOR_CACHE_DIR` and `TRITON_CACHE_DIR` and pre-populating both caches from previous jobs, the compile times can still be > 30 mins.

There are only a handful of input shapes/types that we need to recompile for and these are known ahead of time. As a result, ideally it would be great to `torch.compile` AOT with a single GPU, serialize that model and load it for training on a large number of GPUS without having to incur any compile overhead during training.

**Comment by isidentical** [comment]
This is not only for training, but also for inference workloads where the demand scales quite a bit which causes new machines with the same exact environment + GPUs to come up and down, processing same/similiar input shapes, but each one of those going through a 2min+ compile time for even a small inference pipeline like SDXL (with `TORCHINDUCTOR_CACHE_DIR` + `TORCHINDUCTOR_FX_GRAPH_CACHE` on torch nightlies). 

We've also tried using the `torch.export()` and `torch.aot_compile()` but they both have their own quirks (which in turns creates very hard to trace/reproduce errors) and that makes it super hard to achieve a simple save/load pair for the same exact gpu+environment combination. I'd really love if there was a convient way I can tell `torch.compile()` on a program without any graph breaks to just dump whatever it generated/traced into a cache dir that is only valid for the same GPU+CUDA version+Triton version etc. and only pay the overhead once.

**Comment by Chillee** [comment]
@isidentical I'm curious, what level of compilation times would you find acceptable for your use case?

**Comment by isidentical** [comment]
@Chillee for a hot load on the same exact environment (same GPU, same CUDA version, same Python version, and more importantly same input shapes), ideally, none. For AOT use cases, we have been using TensorRT, and even though it's not 0 overhead when loading an already compiled, it's pretty good. But I'd rather use `torch.compile` with an option to fallback to eager execution if the current execution needs a re-compile operation than use TensorRT on production.

**Comment by jon-chuang** [comment]
I think we can definitely get a beta prototype done very quickly.

It involves cacheing on serializable aspects of a code object. 

I'm inspired by the JIT experienced aimed for in Modular's Mojo.

As secondary goals, we can also benefit from things like deterministic config https://github.com/pytorch/pytorch/pull/111318/, or guarding on codeobject match https://github.com/pytorch/pytorch/pull/110828

If you're ok with a draft prototype (prior to serious design doc/finalization), I can work on one @ezyang @Chillee

**Comment by jon-chuang** [comment]
It would be good to have some profiling to understand what fraction of your compilation process is spent on each stage.

There are many tiers to cacheing the compilation:
1. Guard-level
2. Generated fx <-> backend compiler
3. Backend-level

Each presents its own challenges. Backend-level seems to be partly implemented.

For guard-level cacheing (cacheing at the `torch.compile` interface), (at least) the following are required to be serialized:

1. Runtime wrappers
    - e.g. eager code
    - AOTAutograd runtime wrapper
    - Inductor runtime wrappers
2. Codegen artifact
    - cubin (Triton), .so (cpp)
3. Guards - need to be process independent (e.g. no `ID_MATCH`/`DATA_PTR_MATCH` (e.g. from `mark_static_address` - only used in optimizer, etc.)
    - At most, we cannot be 100% correct, and can guard on things like object name / def lineno / codeobject.

**Comment by jon-chuang** [comment]
I think the safest bet to get this to work well for training is probably:
1. Extending the `aot_compile` approach to GPU .so artifact
2. Extending the `aot_compile` approach to also cache while tracing with Dynamo

This way, while one might retrace the graph module, usually dynamo tracing takes on the upper end ~100+ seconds, then you get to hit the aot_compile backend cache.

Although, I have to say, I find it suspicious that `TORCHINDUCTOR_FX_GRAPH_CACHE` does not help much. That's because it ought to get rid of the inductor side of the compilation entirely.

Does this mean that _all of your overhead is dynamo_?

**Comment by jon-chuang** [comment]
If we can separate out:
1. the backend cacheing with `TORCHINDUCTOR_FX_GRAPH_CACHE`
2. the frontend cacheing with `TORCHDYNAMO_TRACING_CACHE`, which takes in some code object and serializable guards, and spits out an FX graph
    - There is already an idea to carry this out on a smaller scale - cacheing the FX graph for inlined function calls / loop bodies https://github.com/pytorch/pytorch/issues/112303
3. if AOTAutograd is identified as a bottleneck, an `AOT_AUTOGRAD_CACHE`, which takes in some fx graph and generates a runtime wrapper, around a dynamically compiled object (offload to the backend to do its own cacheing)

then we have a multi-tier way to incrementally get rid of overhead, while still allowing recompilation fallback, and not only that, potentially, _shared recompilation cacheing_, while still relying on a lot of runtime infra at each layer of the torch.compile stack.

---

@isidentical would you be able to provide a repro script / steps where I can try to replicate and profile your workload?

**Comment by jon-chuang** [comment]
Just eyeballing, it seems a lot of overhead is indeed coming from AOTAutograd, and even tracing (depends). I suspect AOTAutograd is going to be the bottleneck for training. 

I think it's probably not going to be difficult to implement cacheing for it, but we need to be careful about a lot of things - for example, just the input fx.GraphModule alone is not enough, we also need to guard on the `example_input`s and ambient state (e.g. autocast state, grad mode).

---

Btw, @isidentical, have you made sure to run your inference under `torch.no_grad()`or set your model to `model.eval()`? It will affect the AOTAutograd compilation time I believe.

**Comment by jon-chuang** [comment]
Here is my result from compiling stable diffusion 2:


```
[2023-11-21 00:12:30,439] torch._dynamo.utils: [INFO] TorchDynamo compilation metrics:
[2023-11-21 00:12:30,439] torch._dynamo.utils: [INFO] Function                                  Runtimes (s)
[2023-11-21 00:12:30,439] torch._dynamo.utils: [INFO] --------------------------------------  --------------
[2023-11-21 00:12:30,439] torch._dynamo.utils: [INFO] _compile.<locals>.compile_inner                29.2221
[2023-11-21 00:12:30,439] torch._dynamo.utils: [INFO] OutputGraph.call_user_compiler                 18.5476
[2023-11-21 00:12:30,439] torch._dynamo.utils: [INFO] create_aot_dispatcher_function                 19.8
[2023-11-21 00:12:30,439] torch._dynamo.utils: [INFO] compile_fx.<locals>.fw_compiler_base            6.8342
[2023-11-21 00:12:30,439] torch._dynamo.utils: [INFO] CachingAutotuner.benchmark_all_configs          4.1395

```

As we can see, `create_aot_dispatcher_function` indeed dominates.

**Comment by jon-chuang** [comment]
Here is the rough comparison of time taken between running functionalized_foward and running compiler_fn (aot_dispatch_base):
```
FUNCTIONALIZED FORWARD 0.9658653736114502
COMPILER 2.2161993980407715
```

**Comment by jon-chuang** [comment]
So maybe all we need is make_fx cacheing :)

**Comment by isidentical** [comment]
> @isidentical would you be able to provide a repro script / steps where I can try to replicate and profile your workload?

Definitely! Here is what I were able to isolate from our production code: https://gist.github.com/isidentical/be0edd4c88068759bd220097a571b9b0 (as well as the environment itself, which is based on ~nov 15th torch nightly). 

|                              | timing (s) |
|------------------------------|------------|
| Fresh (no caching)           | 203s       |
| Persistent cache             | 142s       |
| FX graphs + Persistent cache | 71s        |

And this is actually much better than expected (I remember seeing this function taking much longer on some occasions but can't tell if it is because there is some nondeterminism going on in the cache level or sometimes the same cache dir gets overriden maybe or something else), but it is still very noticeable in terms of overhead.


 Will try with the suggestions you have posted as well `torch.no_grad()`/`model.eval()` (diffusers should automatically enable `inference_mode`  when performing the inference but maybe I can give it a even before).

**Comment by jon-chuang** [comment]
> Will try with the suggestions you have posted as well torch.no_grad()/model.eval() (diffusers should automatically enable inference_mode when performing the inference but maybe I can give it a even before).

Yes, it's probably true if you are using huggingface (as it is inference-focused), but something worth trying at least (esp if deploying some other framework's code).

**Comment by pritamdamania87** [comment]
@jon-chuang Here are some metrics from our runs:


```
TorchDynamo compilation metrics:
Function                                   Runtimes (s)
---------------------------------------  --------------
_compile                                       353.797
OutputGraph.call_user_compiler                 289.737
create_aot_dispatcher_function                 289.135
compile_fx.<locals>.fw_compiler                121.575
GraphLowering.run                               35.8149
GraphLowering.compile_to_module                153.171
Scheduler.__init__                              62.2375
Scheduler.codegen                               27.4698
WrapperCodeGen.generate                         37.6984
CachingAutotuner.autotune_to_one_config         23.6054
compile_fx.<locals>.bw_compiler                 69.7454



TorchDynamo compilation metrics:
Function                                   Runtimes (s)
---------------------------------------  --------------
_compile                                       745.292
OutputGraph.call_user_compiler                 593.602
create_aot_dispatcher_function                 592.38
compile_fx.<locals>.fw_compiler                244.366
GraphLowering.run                               73.0489
GraphLowering.compile_to_module                299.881
Scheduler.__init__                             126.119
Scheduler.codegen                               52.6031
WrapperCodeGen.generate                         79.5505
CachingAutotuner.autotune_to_one_config         27.1259
compile_fx.<locals>.bw_compiler                133.357


TorchDynamo compilation metrics:
Function                                   Runtimes (s)
---------------------------------------  --------------
_compile                                      1641.97
OutputGraph.call_user_compiler                1207.88
create_aot_dispatcher_function                1205.63
compile_fx.<locals>.fw_compiler                491.42
GraphLowering.run                              129.163
GraphLowering.compile_to_module               4145.12
Scheduler.__init__                             250.7
Scheduler.codegen                               90.6583
WrapperCodeGen.generate                        180.768
CachingAutotuner.autotune_to_one_config        185.6
compile_fx.<locals>.bw_compiler               3791.39


TorchDynamo compilation metrics:
Function                                   Runtimes (s)
---------------------------------------  --------------
_compile                                      1649.81
OutputGraph.call_user_compiler                1213.43
create_aot_dispatcher_function                1211.13
compile_fx.<locals>.fw_compiler                493.741
GraphLowering.run                              129.971
GraphLowering.compile_to_module               4147.59
Scheduler.__init__                             251.757
Scheduler.codegen                               91.1865
WrapperCodeGen.generate                        180.841
CachingAutotuner.autotune_to_one_config        185.6
compile_fx.<locals>.bw_compiler               3792.4


TorchDynamo compilation metrics:
Function                                   Runtimes (s)
---------------------------------------  --------------
_compile                                       2041.35
OutputGraph.call_user_compiler                 1514.86
create_aot_dispatcher_function                 1511.88
compile_fx.<locals>.fw_compiler                 608.982
GraphLowering.run                               167.213
GraphLowering.compile_to_module                4280.19
Scheduler.__init__                              314.94
Scheduler.codegen                               116.429
WrapperCodeGen.generate                         224.092
CachingAutotuner.autotune_to_one_config         185.6
compile_fx.<locals>.bw_compiler                3849.51
```

**Comment by jon-chuang** [comment]
@pritamdamania87 do you also have similar numbers for `TORCHINDUCTOR_FX_GRAPH_CACHE=1` or was this with that setting already enabled?

**Comment by pritamdamania87** [comment]
> @pritamdamania87 do you also have similar numbers for TORCHINDUCTOR_FX_GRAPH_CACHE=1 or was this with that setting already enabled?

This was without that flag, with `TORCHINDUCTOR_FX_GRAPH_CACHE=1` seems to improve things by 10-15%

**Comment by masnesral** [comment]
Just FYI @jamesjwu based on some of the comments here it sounds like the AOTAutograd cache will be welcomed here.

**Comment by yf225** [comment]
cc. @oulgen @zou3519 this might be relevant for potential AOT compile work in H1 roadmap

**Comment by ezyang** [comment]
Check out https://github.com/pytorch/pytorch/pull/141700

**Comment by bobrenjc93** [comment]
This is now possible today with precompile on one rank: https://github.com/pytorch/torchtitan/tree/main/torchtitan/experiments/graph_trainer#pre-compile-compile-on-one-rank
